### ISSUE TITLE
caffe2 conversion fails on super resolution example

### ISSUE THREAD

**Issue by xonobo** [pytorch/pytorch#18429]
## To Reproduce

```
python3 -m venv pytorch
source pytorch/bin/activate
pip3 install https://download.pytorch.org/whl/cu100/torch-1.0.1.post2-cp36-cp36m-linux_x86_64.whl
pip3 install torchvision
pip3 install onnx
pip3 install future
wget https://pytorch.org/tutorials/_downloads/8ef9234cc18eea8ec4c2dd6046887cf8/super_resolution_with_caffe2.py
python super_resolution_with_caffe2.py
```

produces:

```
Mismatch: 99.9%
Max absolute difference: 6.1213245
Max relative difference: 994661.1
 x: array([[[[ 2.394e-01,  3.654e-01,  4.718e-01, ...,  1.865e-01,
           3.322e-01,  3.623e-01],
         [ 2.033e-01,  4.136e-01,  5.250e-01, ...,  4.200e-01,...
 y: array([[[[ 0.239,  0.44 ,  0.396, ...,  0.793, -0.133,  0.362],
         [ 0.056, -0.747, -0.327, ...,  1.044,  0.898,  0.873],
         [ 0.252,  1.514,  0.212, ..., -0.593,  0.011, -0.107],...
```

**Comment by soumith** [comment]
cc: @houseroad @zrphercule

**Comment by xonobo** [comment]
Similar results for the nightly build pips.

**Comment by ezyang** [comment]
Super resolution is broken. It's been disabled for a while in CI, and in fact the test is failing when I run it locally.

CUDA:

```
======================================================================
FAIL: test_super_resolution (__main__.TestModels)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/scratch/ezyang/pytorch-tmp/test/common_utils.py", line 159, in wrapper
    fn(*args, **kwargs)
  File "test/onnx/test_models.py", line 76, in test_super_resolution
    self.exportTest(toC(SuperResolutionNet(upscale_factor=3)), toC(x), atol=1e-6)
  File "test/onnx/test_models.py", line 42, in exportTest
    verify(model, inputs, backend, rtol=rtol, atol=atol)
  File "/scratch/ezyang/pytorch-tmp/test/onnx/verify.py", line 454, in verify
    run_helper(torch_out, args)
  File "/scratch/ezyang/pytorch-tmp/test/onnx/verify.py", line 452, in run_helper
    errs.checkAlmostEqual(x.data.cpu().numpy(), y, "In output {}".format(i))
  File "/scratch/ezyang/pytorch-tmp/test/onnx/verify.py", line 60, in checkAlmostEqual
    self.almostEqualAndThen(x, y, msg, self.addErr)
  File "/scratch/ezyang/pytorch-tmp/test/onnx/verify.py", line 72, in almostEqualAndThen
    equal_nan=True, verbose=True)
  File "/scratch/ezyang/pytorch-tmp-env/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 1501, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/scratch/ezyang/pytorch-tmp-env/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 827, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0.01, atol=1e-06

Mismatch: 90.1%
Max absolute difference: 2.7199602
Max relative difference: 701.5824
 x: array([[[[ 0.269054,  0.291963,  0.318168, ..., -0.218563, -0.037737,
          -0.069209],
         [-0.030326, -0.19607 ,  0.155896, ...,  0.326319, -0.41674 ,...
 y: array([[[[ 0.269054,  0.275228,  0.225867, ...,  0.210519,  0.122266,
          -0.069209],
         [ 0.388265,  0.532256,  0.546107, ...,  0.03942 ,  0.026328,...

----------------------------------------------------------------------
Ran 1 test in 2.893s

FAILED (failures=1)
```

CPU:

```
diff --git a/test/onnx/test_models.py b/test/onnx/test_models.py
index 04aca5aa2..4966c6b0e 100644
--- a/test/onnx/test_models.py
+++ b/test/onnx/test_models.py
@@ -73,7 +73,7 @@ class TestModels(TestCase):
         x = Variable(
             torch.randn(BATCH_SIZE, 1, 224, 224).fill_(1.0)
         )
-        self.exportTest(toC(SuperResolutionNet(upscale_factor=3)), toC(x), atol=1e-6)
+        self.exportTest(SuperResolutionNet(upscale_factor=3), x, atol=1e-6)
 
     def test_alexnet(self):
         x = Variable(
```

then:

```
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/scratch/ezyang/pytorch-tmp-env/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 827, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0.01, atol=1e-06

Mismatch: 90.1%
Max absolute difference: 2.7199597
Max relative difference: 701.58246
 x: array([[[[ 0.269054,  0.291962,  0.318168, ..., -0.218563, -0.037737,
          -0.069209],
         [-0.030326, -0.19607 ,  0.155896, ...,  0.326319, -0.41674 ,...
 y: array([[[[ 0.269054,  0.275228,  0.225867, ...,  0.210519,  0.122266,
          -0.069209],
         [ 0.388265,  0.532256,  0.546107, ...,  0.03942 ,  0.026328,...

----------------------------------------------------------------------
Ran 1 test in 0.757s

FAILED (failures=1)
```

**Comment by geoffsmith** [comment]
Does anyone know what might be causing the issue? I've built my own super-resolution model and am trying to get ONNX to work. It's exporting fine, but re-loading the model and running inferences gives completely different results - I'm wondering if this is the same issue I'm getting. Maybe something to do with exporting pixel shuffle to onnx?

**Comment by ezyang** [comment]
A while ago, we had some problems with certain models because of numeric
stability issues: there were small numerical differences between PyTorch
and Caffe2 and the model architecture amplified these differences.

The easiest way to debug this problem is to do a layer-by-layer
comparison, and see where the divergence happens.  Let us know if you
find out.

Excerpts from Geoff Smith's message of 2019-06-06 12:49:16 -0700:
> Does anyone know what might be causing the issue? I've built my own super-resolution model and am trying to get ONNX to work. It's exporting fine, but re-loading the model and running inferences gives completely different results - I'm wondering if this is the same issue I'm getting. Maybe something to do with exporting pixel shuffle to onnx?
>

**Comment by geoffsmith** [comment]
I'm pretty sure now there's an issue with exporting PixelShuffle to ONNX. I've made a minimal script to demonstate: https://gist.github.com/geoffsmith/0b69070e96262dffe4ae1c22e9126bf3

I think that the PixelShuffle ONNX export function re-arranging the tensors slightly differently to the way that PyTorch does it. In the script I show that the values in a simple (1, 1, 4, 4) tensor are ordered differently when a PixelShuffle module is exported and then re-imported.

In the script, I've also made an alternative PixelShuffle that was copied off the C++ pixel_shuffle function, but using Python. This gives the same output as PixelShuffle, but also works correctly when exported to ONNX and re-imported back again.

This will defo break the above model since it uses PixelShuffle (though of course there could also be numerical stability issues on top of this :)
